README.md
# Subscription service
[![Run Maven Tests](https://github.com/qbicsoftware/subscription-service/actions/workflows/run_tests.yml/badge.svg?branch=main)](https://github.com/qbicsoftware/subscription-service/actions/workflows/run_tests.yml)
[![CodeQL](https://github.com/qbicsoftware/subscription-service/actions/workflows/codeql-analysis.yml/badge.svg?branch=main)](https://github.com/qbicsoftware/subscription-service/actions/workflows/codeql-analysis.yml)

Small REST service to handle sample status tracking subscription cancellations.
This service allows to securely generate cancellation tokens and triggers cancellation of subscriptions

## Endpoints

### POST /subscriptions/tokens

To cancel a subscription a token needs to be generated in order to allow unsubscription.
The cancellation token is returned as String.

To generate a token a valid cancellation request needs to be send:
```
{
    "project": "string",
    "userId": "string"
}
```

A valid request triggers the generation of a token based on the cancellation request.
The service answers with 201 and the token in plain text:

```
For_lfbnS9iTi4Nmwnei4LA_f8SHga1Rdz4yw6aT8zz0V8PaHm1QEbKQTv1jGCEA
```

Invalid requests are plain text requests. The service answers with 400 or 401.

### DELETE subscriptions/{token}
Cancels a subscription for the given token. The token encodes the project and the user id for the unsubscription.
A token can look like this:
```
For_lfbnS9iTi4Nmwnei4LA_f8SHga1Rdz4yw6aT8zz0V8PaHm1QEbKQTv1jGCEA
```
If the unsubscription was successful the service answers with 204 and no content.
Unsuccessful removal of a subscription, e.g. due to invalid tokens, result answer with 400. 
In case of correctly formatted tokens but unsuccessfull unsubscription, the answer will be 422 (Unprocessable Entity).

## Env vars
The env variables contain information about the salt and the secret. Both of them are used to encrypt and decrypt user information.

| environment variable       | description               |
|----------------------------|---------------------------|
| `USER_DB_DIALECT`          | The database dialect      |
| `USER_DB_DRIVER`           | The database driver       |
| `USER_DB_URL`              | The database host address |
| `USER_DB_USER_NAME`        | The database user name    |
| `USER_DB_USER_PW`          | The database password     |
| `ENCRYPTION_KEY`           | The encryption key        |
| `ENCRYPTION_SALT`          | The encryption salt       |
| `SERVICE_OFFICER_NAME`     | The office name           |
| `SERVICE_OFFICER_PASSWORD` | The officer password      |

The application properties file could look like the following:
```
databases.users.database.dialect=${:org.hibernate.dialect.MariaDBDialect}
databases.users.database.driver=${USER_DB_DRIVER:com.mysql.cj.jdbc.Driver}
databases.users.database.url=${USER_DB_URL:localhost}
databases.users.user.name=${USER_DB_USER_NAME:myusername}
databases.users.user.password=${USER_DB_USER_PW:astrongpassphrase!}
encryption.secret=${ENCRYPTION_KEY:12345678901234}
encryption.salt=${ENCRYPTION_SALT:123456789!1234567}
service.officer.name=${SERVICE_OFFICER_NAME:ChuckNorris}
service.officer.password=${SERVICE_OFFICER_PASSWORD:astrongpassphrase!}
```

## How to start the service 
To run the service you first need to package it. You need Java 17 to do that.

```
mvn clean package
```

Deploy the generated jar and start the service.




CHANGELOG.md
# Changelog

# 1.0.0-rc.3

* Change API endpoints
  * `DELETE /subscriptions/{token}` to remove a subscription
  * `POST /subscriptions/tokens` to create an unsubscription token

# 1.0.0-rc.2

* Disabled CSRF, until we have a working implementation for the clients

# 1.0.0-rc.1 

* Added first REST endpoints to enable 
  * requests to retreive a cancellation token and
  * submit the subscription cancellation


src/main/java/life/qbic/subscriptions/SubscriptionController.java
package life.qbic.subscriptions;

import static org.slf4j.LoggerFactory.getLogger;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.enums.SecuritySchemeType;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.security.SecurityScheme;
import io.swagger.v3.oas.annotations.tags.Tag;
import life.qbic.subscriptions.encryption.DecryptionException;
import life.qbic.subscriptions.encryption.EncryptionException;
import life.qbic.subscriptions.encryption.RequestDecrypter;
import life.qbic.subscriptions.encryption.RequestEncrypter;
import life.qbic.subscriptions.subscriptions.CancellationRequest;
import life.qbic.subscriptions.subscriptions.SubscriptionRepository;
import org.slf4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@Tag(name = "Subscription", description = "Subscription API")
@SecurityScheme(name = "basic", type = SecuritySchemeType.HTTP, scheme = "basic")
@RestController
@RequestMapping("/subscriptions")
public class SubscriptionController {

  private static final Logger log = getLogger(SubscriptionController.class);

  RequestEncrypter requestEncrypter;

  RequestDecrypter requestDecrypter;

  SubscriptionRepository subscriptionRepository;

  @Autowired
  SubscriptionController(
      SubscriptionRepository subscriptionRepository,
      RequestDecrypter requestDecrypter,
      RequestEncrypter requestEncrypter) {
    this.subscriptionRepository = subscriptionRepository;
    this.requestDecrypter = requestDecrypter;
    this.requestEncrypter = requestEncrypter;
  }

  @Operation(summary = "Cancel a subscription",
      parameters = {
          @Parameter(name = "token",
              description = "The token of an encrypted cancel request.",
              example = "For_lfbnS9iTi4Nmwnei4LA_f8SHga1Rdz4yw6aT8zz0V8PaHm1QEbKQTv1jGCEA", schema = @Schema(implementation = String.class))
      },
      responses = {
          @ApiResponse(responseCode = "204", description = "No content. Your subscription was cancelled."
          ),
          @ApiResponse(responseCode = "400", description = "Bad request. Your cancellation request was not successful.",
              content = @Content(mediaType = "text/plain")
          ),
          @ApiResponse(responseCode = "422", description = "Unprocessable entity. Your cancellation request was not successful.",
              content = @Content(mediaType = "text/plain")
          )
      })
  @SecurityRequirement(name = "basic")
  @DeleteMapping(value = "/{token}")
  @ResponseStatus(value = HttpStatus.NO_CONTENT)
  public void cancelSubscriptionByToken(@PathVariable String token) {
    var cancellationRequest = requestDecrypter.decryptCancellationRequest(token);
    removeSubscription(cancellationRequest);
  }

  @Operation(summary = "Request a subscription cancel token",
      responses = {
          @ApiResponse(responseCode = "201", description = "Subscription cancel token was created.",
              content = @Content(mediaType = "text/plain", schema = @Schema(example = "For_lfbnS9iTi4Nmwnei4LA_f8SHga1Rdz4yw6aT8zz0V8PaHm1QEbKQTv1jGCEA"))
          ),
          @ApiResponse(responseCode = "400", description = "Bad request. Your cancellation request might not be correct.",
              content = @Content(mediaType = "text/plain")
          ),
          @ApiResponse(responseCode = "401", description = "Full authentication required.",
              content = @Content(mediaType = "text/plain")
          )
      })
  @SecurityRequirement(name = "basic")
  @PostMapping(value = "/tokens")
  public ResponseEntity<String> generateToken(@RequestBody CancellationRequest cancellationRequest) {
    validateRequest(cancellationRequest);
    var cancellationRequestToken = requestEncrypter.encryptCancellationRequest(cancellationRequest);
    return new ResponseEntity<>(cancellationRequestToken, HttpStatus.CREATED);
  }

  private void removeSubscription(CancellationRequest request) {
    try {
      subscriptionRepository.cancelSubscription(request);
    } catch (Exception e) {
      log.error(e.getMessage(), e);
      throw new CancellationFailure("Unexpected failure.");
    }
  }

  private void validateRequest(CancellationRequest cancellationRequest) {
    var project = cancellationRequest.project();
    var userId = cancellationRequest.userId();
    if (project == null || project.isBlank()) {
      throw new MissingPropertyException("project");
    }
    if (userId == null || userId.isBlank()) {
      throw new MissingPropertyException("userId");
    }
  }

  @ExceptionHandler({EncryptionException.class})
  private ResponseEntity<String> encryptionException() {
    return new ResponseEntity<>(
        "Cancellation request generation failed", HttpStatus.BAD_REQUEST);
  }

  @ExceptionHandler({DecryptionException.class})
  private ResponseEntity<String> decryptionException() {
    return new ResponseEntity<>(
        "Decryption of your request has failed", HttpStatus.BAD_REQUEST);
  }

  @ExceptionHandler({CancellationFailure.class})
  private ResponseEntity<String> cancellationFailure(CancellationFailure e) {
    return new ResponseEntity<>(
        "Subscription could not be cancelled. Reason: " + e.reason, HttpStatus.UNPROCESSABLE_ENTITY);
  }

  @ExceptionHandler({MissingPropertyException.class})
  private ResponseEntity<String> missingProperty(MissingPropertyException e) {
    return new ResponseEntity<>(
        "Missing content for required property: " + e.missingProperty, HttpStatus.BAD_REQUEST);
  }

  /**
   * Helper exception class, to indicate failing subscription cancellations requests
   */
  public static class CancellationFailure extends RuntimeException {

    final String reason;

    public CancellationFailure(String reason) {
      super();
      this.reason = reason;
    }
  }

  /**
   * Helper exception class, to indicate missing request properties
   */
  public static class MissingPropertyException extends RuntimeException {

    final String missingProperty;

    public MissingPropertyException(String missingProperty) {
      super("Missing property '" + missingProperty + "'");
      this.missingProperty = missingProperty;
    }
  }

}
